### ISSUE TITLE
[FEATURE-REQUEST/ HELP REQUIRED] Getting first and last members of a group

### ISSUE THREAD

**Issue by Chris-hughes10** [vaexio/vaex#1973]
**Description**

Hi, I would like to be able to retrieve the first and last members of a group after a groupby operation, which is straightforward to do  using pandas, but I seem to be unable to do this using Vaex. Looking through the docs, I have encountered the aggregators `vaex.agg.first` and `vaex.agg.last` which appeared as though they would solve this problem, but I can't get these to work at all. However, this may be user error, as I can't find documentation or examples detailing their usage anywhere. 

As a motivating example, consider the following DataFrame:

![image](https://user-images.githubusercontent.com/31883449/158644794-d14b7de5-0765-4939-b315-996e8631c05e.png)

I would like to retrieve the content for the first and last rows in each 'Class'. To get the first item, I would expect that you should be able to do something along the lines of:

`df.groupby('Class').agg({'content': vaex.agg.first(df.content, df.idx)})`

or

`df.groupby('Class', agg=vaex.agg.first('content', 'idx'))`

unfortunately I can't get anything like this to work.

The version that I am using is
```
{'vaex': '4.8.0',
 'vaex-core': '4.8.0',
 'vaex-viz': '0.5.1',
 'vaex-hdf5': '0.12.0',
 'vaex-server': '0.8.1',
 'vaex-astro': '0.9.0',
 'vaex-jupyter': '0.7.0',
 'vaex-ml': '0.17.0'}
```
If this is possible, any help would be much appreciated, and an example should be added to the docs. Alternatively, I believe that this would be useful functionality to implement.

Thanks

**Comment by Chris-hughes10** [comment]
Apologies if my issue came across as unhelpful, I thought that I would keep it concise in case there was a quick fix that could be pointed out, if this was documented elsewhere, or if I was doing something dumb; I am happy to provide more details. 

The df here is a simplified example of the problem I am actually working on, so I generated it as follows:
```
import pandas as pd
import vaex

df = pd.DataFrame([{'Class': 'A', 'idx': 0, 'content': 'xxx'},
                  {'Class': 'A', 'idx': 1, 'content': 'xxx'},
                  {'Class': 'A', 'idx': 2, 'content': 'xxx'},
                  {'Class': 'A', 'idx': 3, 'content': 'xxx'},
                  {'Class': 'A', 'idx': 4, 'content': 'xxx'},
                  {'Class': 'A', 'idx': 5, 'content': 'xxx'},
                   {'Class': 'B', 'idx': 0, 'content': 'xxx'},
                  {'Class': 'B', 'idx': 1, 'content': 'xxx'},
                  {'Class': 'B', 'idx': 2, 'content': 'xxx'},
                  {'Class': 'B', 'idx': 3, 'content': 'xxx'},
                   
                  ])

df = vaex.from_pandas(df)
```

Interestingly I get a few different errors depending on the approach:

`df.groupby('Class').agg({'content': vaex.agg.first('content', 'idx')})`

```
ValueError                                Traceback (most recent call last)
Input In [9], in <module>
----> 1 df.groupby('Class').agg({'content': vaex.agg.first('content', 'idx')})

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:922, in GroupBy.agg(self, actions, delay, progress)
    920     return result
    921 else:
--> 922     self.df.execute()
    923     return result.get()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:412, in DataFrame.execute(self)
    410         print(repr(task))
    411 if self.executor.tasks:
--> 412     self.executor.execute()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:308, in ExecutorLocal.execute(self)
    307 def execute(self):
--> 308     for _ in self.execute_generator():
    309         pass

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:401, in ExecutorLocal.execute_generator(self, use_async)
    395     return task_part
    396 # We want at least 1 task part (otherwise we cannot do any work)
    397 # then we ask for the task part how often we should split
    398 # This means that we can have 100 threads, but only 2 task parts
    399 # In this case, evaluation of expressions is still multithreaded,
    400 # but aggregation is reduced to effectively 2 threads.
--> 401 task_part_0 = create_task_part()
    402 ideal_task_splits = task_part_0.ideal_splits(self.thread_pool.nthreads)
    403 assert ideal_task_splits <= self.thread_pool.nthreads, f'Cannot have more splits {ideal_task_splits} then threads {self.thread_pool.nthreads}'

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:389, in ExecutorLocal.execute_generator.<locals>.create_task_part()
    387 def create_task_part():
    388     nonlocal memory_usage
--> 389     task_part = encoding.decode('task-part-cpu', spec, df=task.df, nthreads=nthreads)
    390     memory_usage += task_part.memory_usage()
    391     for task_checker in task_checkers:

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/encoding.py:384, in Encoding.decode(self, typename, value, **kwargs)
    383 def decode(self, typename, value, **kwargs):
--> 384     decoded = self.registry[typename].decode(self, value, **kwargs)
    385     return decoded

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/encoding.py:51, in make_class_registery.<locals>.encoding.decode(encoding, spec, **kwargs)
     49 type = spec.pop(f'{groupname}-type')
     50 cls = _encoding_types[type]
---> 51 return cls.decode(encoding, spec, **kwargs)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/cpu.py:667, in TaskPartAggregation.decode(cls, encoding, spec, df, nthreads)
    665 for agg in aggs:
    666     agg._prepare_types(df)
--> 667 return cls(df, grid, aggs, dtypes, initial_values=values, nthreads=nthreads)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/cpu.py:514, in TaskPartAggregation.__init__(self, df, binners, aggregation_descriptions, dtypes, initial_values, nthreads)
    512 selections = _ensure_list(selection)
    513 initial_values_i = initial_values[i] if initial_values else None
--> 514 self.aggregations.append((aggregator_descriptor, selections, list(create_aggregator(aggregator_descriptor, selections, initial_values_i)), selection_waslist))

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/cpu.py:499, in TaskPartAggregation.__init__.<locals>.create_aggregator(aggregator_descriptor, selections, initial_values)
    496 def create_aggregator(aggregator_descriptor, selections, initial_values):
    497     # for each selection, we have a separate aggregator, sharing the grid and binners
    498     for i, selection in enumerate(selections):
--> 499         agg = aggregator_descriptor._create_operation(self.grid, nthreads)
    500         self.nbytes += sys.getsizeof(agg)
    501         if initial_values is not None:

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/agg.py:278, in AggregatorDescriptorBasic._create_operation(self, grid, nthreads)
    277 def _create_operation(self, grid, nthreads):
--> 278     agg_op_type = vaex.utils.find_type_from_dtype(vaex.superagg, self.name + "_", self.dtype_in)
    279     bytes_per_cell = self.dtype_out.numpy.itemsize
    280     cells = reduce(operator.mul, [len(binner) for binner in grid.binners], 1)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/utils.py:767, in find_type_from_dtype(namespace, prefix, dtype, transient, support_non_native)
    765     return getattr(namespace, name)
    766 else:
--> 767     raise ValueError('Could not find a class (%s), seems %s is not supported' % (name, dtype))

ValueError: Could not find a class (AggFirst_string), seems string is not supported
```

`df.groupby('Class').agg({'content': vaex.agg.first(df.content, df.idx)})`

```
TypeError                                 Traceback (most recent call last)
Input In [11], in <module>
----> 1 df.groupby('Class').agg({'content': vaex.agg.first(df.content, df.idx)})

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:922, in GroupBy.agg(self, actions, delay, progress)
    920     return result
    921 else:
--> 922     self.df.execute()
    923     return result.get()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:412, in DataFrame.execute(self)
    410         print(repr(task))
    411 if self.executor.tasks:
--> 412     self.executor.execute()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:308, in ExecutorLocal.execute(self)
    307 def execute(self):
--> 308     for _ in self.execute_generator():
    309         pass

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:345, in ExecutorLocal.execute_generator(self, use_async)
    343 if not tasks:
    344     break
--> 345 tasks = _merge(tasks)
    346 run = Run(tasks)
    347 self.passes += 1

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:137, in _merge(tasks)
    135 for df in dfs:
    136     tasks_df = [k for k in tasks if k.df == df]
--> 137     tasks_merged.extend(_merge_tasks_for_df(tasks_df, df))
    138 return tasks_merged

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/execution.py:161, in _merge_tasks_for_df(tasks, df)
    159     return result
    160 task_merged.signal_progress.connect(progress_wrapper)
--> 161 task_merged.add_aggregation_operation(subtask.aggregation_description)
    162 @vaex.delayed
    163 def assign(value, i=i, subtask=subtask):
    164     subtask.fulfill(value[i])

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/tasks.py:492, in TaskAggregations.add_aggregation_operation(self, aggregator_descriptor)
    490 self.expressions_all.extend(aggregator_descriptor.expressions)
    491 # TODO: optimize/remove?
--> 492 self.dtypes = {expr: self.df.data_type(expr).index_type for expr in self.expressions_all}
    493 return task

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/tasks.py:492, in <dictcomp>(.0)
    490 self.expressions_all.extend(aggregator_descriptor.expressions)
    491 # TODO: optimize/remove?
--> 492 self.dtypes = {expr: self.df.data_type(expr).index_type for expr in self.expressions_all}
    493 return task

TypeError: unhashable type: 'Expression'
```

`df.groupby('Class', agg=vaex.agg.first(df.content, df.idx))`

```
AttributeError                            Traceback (most recent call last)
Input In [8], in <module>
----> 1 df.groupby('Class', agg=vaex.agg.first(df.content, df.idx))

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:7000, in DataFrameLocal.groupby(self, by, agg, sort, assume_sparse, row_limit, copy, progress, delay)
   6998     else:
   6999         return groupby.agg(agg, delay=delay, progress=progressbar_agg)
-> 7000 return self._delay(delay, progressbar.exit_on(next(groupby._promise_by)))

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:1689, in DataFrame._delay(self, delay, task, progressbar)
   1687 else:
   1688     self.execute()
-> 1689     return task.get()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/aplus/__init__.py:170, in Promise.get(self, timeout)
    168     return self._value
    169 else:
--> 170     raise self._reason

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:121, in Promise.then.<locals>.callAndReject(r)
    119 try:
    120     if aplus._isFunction(failure):
--> 121         ret.fulfill(failure(r))
    122     else:
    123         ret.reject(r)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/progress.py:91, in ProgressTree.exit_on.<locals>.error(arg)
     89 def error(arg):
     90     self.exit()
---> 91     raise arg

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:106, in Promise.then.<locals>.callAndFulfill(v)
    104 try:
    105     if aplus._isFunction(success):
--> 106         ret.fulfill(success(v))
    107     else:
    108         ret.fulfill(v)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/delayed.py:82, in delayed.<locals>.wrapped.<locals>.call(_)
     80 kwargs_real = {key: promise.get() for key, promise in key_promise}
     81 args_real = list([promise.get() for promise in arg_promises])
---> 82 return f(*args_real, **kwargs_real)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:6999, in DataFrameLocal.groupby.<locals>.next(_ignore)
   6997     return groupby
   6998 else:
-> 6999     return groupby.agg(agg, delay=delay, progress=progressbar_agg)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:923, in GroupBy.agg(self, actions, delay, progress)
    921 else:
    922     self.df.execute()
--> 923     return result.get()

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/aplus/__init__.py:170, in Promise.get(self, timeout)
    168     return self._value
    169 else:
--> 170     raise self._reason

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:121, in Promise.then.<locals>.callAndReject(r)
    119 try:
    120     if aplus._isFunction(failure):
--> 121         ret.fulfill(failure(r))
    122     else:
    123         ret.reject(r)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/progress.py:91, in ProgressTree.exit_on.<locals>.error(arg)
     89 def error(arg):
     90     self.exit()
---> 91     raise arg

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:121, in Promise.then.<locals>.callAndReject(r)
    119 try:
    120     if aplus._isFunction(failure):
--> 121         ret.fulfill(failure(r))
    122     else:
    123         ret.reject(r)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/delayed.py:38, in _log_error.<locals>._wrapped(exc)
     35     print(f"*** DEBUG: Error from {name}", exc)
     36 # import vaex
     37 # vaex.utils.print_stack_trace()
---> 38 raise exc

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:121, in Promise.then.<locals>.callAndReject(r)
    119 try:
    120     if aplus._isFunction(failure):
--> 121         ret.fulfill(failure(r))
    122     else:
    123         ret.reject(r)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/delayed.py:38, in _log_error.<locals>._wrapped(exc)
     35     print(f"*** DEBUG: Error from {name}", exc)
     36 # import vaex
     37 # vaex.utils.print_stack_trace()
---> 38 raise exc

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/promise.py:106, in Promise.then.<locals>.callAndFulfill(v)
    104 try:
    105     if aplus._isFunction(success):
--> 106         ret.fulfill(success(v))
    107     else:
    108         ret.fulfill(v)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/delayed.py:82, in delayed.<locals>.wrapped.<locals>.call(_)
     80 kwargs_real = {key: promise.get() for key, promise in key_promise}
     81 args_real = list([promise.get() for promise in arg_promises])
---> 82 return f(*args_real, **kwargs_real)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:839, in GroupBy.agg.<locals>.aggregate(promise_by)
    837 @vaex.delayed
    838 def aggregate(promise_by):
--> 839     arrays = super(GroupBy, self)._agg(actions, progressbar)
    840     # we don't want non-existing pairs (e.g. Amsterdam in France does not exist)
    841     # but also, e.g. GrouperInteger will always expect missing values
    842     # but they may not aways exist
    843     counts = self.counts

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:707, in GroupByBase._agg(self, actions, progressbar_agg)
    705                     add(aggregate(name), name, override_name=override_name)
    706             else:
--> 707                 add(aggregate, name, override_name=override_name)
    708 return grids

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/groupby.py:665, in GroupByBase._agg.<locals>.add(aggregate, column_name, override_name)
    663 def add(aggregate, column_name=None, override_name=None):
    664     if column_name is None or override_name is not None:
--> 665         column_name = aggregate.pretty_name(override_name, df)
    666     aggregate.edges = True  # is this ok to override?
    667     values = df._agg(aggregate, self.binners, delay=_USE_DELAY, progress=progressbar_agg)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/agg.py:63, in AggregatorDescriptor.pretty_name(self, id, df)
     61 def pretty_name(self, id, df):
     62     if id is None:
---> 63         id = "_".join(map(lambda k: df[k]._label, self.expressions))
     64     return '{0}_{1}'.format(id, self.short_name)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/agg.py:63, in AggregatorDescriptor.pretty_name.<locals>.<lambda>(k)
     61 def pretty_name(self, id, df):
     62     if id is None:
---> 63         id = "_".join(map(lambda k: df[k]._label, self.expressions))
     64     return '{0}_{1}'.format(id, self.short_name)

File /anaconda/envs/azureml_py38/lib/python3.8/site-packages/vaex/dataframe.py:279, in DataFrame.__getattr__(self, name)
    277     return self.__hidden__[name].__get__(self)
    278 else:
--> 279     return object.__getattribute__(self, name)

AttributeError: 'DataFrameLocal' object has no attribute '_label'
```

**Comment by Chris-hughes10** [comment]
When I try to use `vaex.agg.last`, I get the following error: `AttributeError: module 'vaex.agg' has no attribute 'last'`. I have double checked my version, and the docs that I am looking at are 4.8.0, which is consistent with what is installed.

**Comment by yohplala** [comment]
> When I try to use `vaex.agg.last`, I get the following error: `AttributeError: module 'vaex.agg' has no attribute 'last'`. I have double checked my version, and the [docs that I am looking at](https://vaex.io/docs/api.html#vaex.agg.last) are 4.8.0, which is consistent with what is installed.

You need to install from source, not from conda or mamba. The PR has been merged just recently. Did you install from source?

**Comment by JovanVeljanoski** [comment]
I am not 100% sure, but i think the docs follow the latest master.

Also just fyi there are a couple of known issues related to the `first` & `last` aggregators:
https://github.com/vaexio/vaex/pull/1962

**Comment by maartenbreddels** [comment]
> I am not 100% sure, but i think the docs follow the latest master.

yes

And this should be fixed now (and available in the next release (4.9)
